staging/src/k8s.io/apimachinery/pkg/apis/meta/internalversion/types.go

```diff
@@ -66,6 +66,21 @@ type ListOptions struct {
 	// it does not recognize and will return a 410 error if the token can no longer be used because
 	// it has expired.
 	Continue string
+
+	// SendInitialEvents when set together with Watch option, begins the watch stream with synthetic "Added" events of all resources.
+	// Followed by a synthetic "Bookmark" event containing a ResourceVersion after which the server continues streaming events.
+	// This option is meant to be used with the resourceVersion parameter with the following semantics:
+	//
+	// - sendInitialEvent and RV>0 - provides the initial events and when we reach the desired RV,
+	//   the synthetic bookmark event. Note that the desired RV must be >= the current revision the server knows about.
```

Comment:
how does the client distinguish between the bookmark indicating "initial events are done" and the bookmark indicating "server is shutting down"?

Reply:
That's a good question.
The answer is that (at least by design) we will just not send the bookmark indicating "server shutting down" if we didn't already sync. I'm fairly sure the code works fine now, but we should add a test for that.

FWIW - we still don't really send bookmarks on shutdown though - we send them on "watch shutdown" only. But the question remains valid for that usecase (and the answer remains too).

Reply:
in that special state, we won't send a bookmark that is < than the desired one even if the watcher is about to shut down.  I'll add a unit test to cover this case.